internal/proxy/origins/prometheus/matrix.go

```diff
@@ -137,23 +167,38 @@ func (me *MatrixEnvelope) CropToSize(sz int, t time.Time, lur timeseries.Extent)
 
 	for _, s := range me.Data.Result {
 		tmp := s.Values[:0]
+		wg := sync.WaitGroup{}
+		mtx := sync.Mutex{}
 		for _, r := range s.Values {
-			t = r.Timestamp.Time()
-			if _, ok := removals[t]; !ok {
-				tmp = append(tmp, r)
-			}
+			wg.Add(1)
+			go func(p model.SamplePair) {
+				mtx.Lock()
+				t = p.Timestamp.Time()
+				if _, ok := removals[t]; !ok {
+					tmp = append(tmp, p)
+				}
+				mtx.Unlock()
+				wg.Done()
+			}(r)
 		}
+		wg.Wait()
 		s.Values = tmp
 	}
 
 	tl := times.FromMap(removals)
 	sort.Sort(tl)
 	for _, t := range tl {
+		wg := sync.WaitGroup{}
 		for i, e := range el {
-			if e.StartsAt(t) {
-				el[i].Start = e.Start.Add(me.StepDuration)
-			}
+			wg.Add(1)
+			go func(j int, f *timeseries.Extent) {
+				if f.StartsAt(t) {
+					el[j].Start = f.Start.Add(me.StepDuration)
```

Comment:
We may want to add locks to not have concurrent access issues around `el` per each `tl`.

Reply:
Looks good now.